### PR TITLE
Exclude comm handshake from connect timeout

### PR DESCRIPTION
**Pull request by fjetter**
There is not a very specific issue this is fixing but I've come across many connection timeouts over the time and some of them are actually caused by a timed out handshake.

There is no reason for the handshake to be included in the timeout. The timeout is primarily there to ensure that the remote actually exists and we're not attempting to connect to an otherwise dead remote.

Particularly in situations when the event loop is blocked, the handshake can actually take quite a while, particularly if read/write is done sequentially as on main. However, once the connection is established, the entire operation is mostly about sending/receiving some ordinary network messages and we're not timing out such sendrcv operations anywhere else. I believe it is very generally true that the handshake is typically not slow due to network but only due to server load.

### CODE REVIEW

File: distributed/comm/core.py [hunk 1]
```diff
@@ -220,20 +219,8 @@ async def on_connection(
...
-            await wait_for(comm.write(local_info), timeout=timeout)
-            handshake = await wait_for(comm.read(), timeout=timeout)
-            # This would be better, but connections leak if worker is closed quickly
-            # write, handshake = await asyncio.gather(comm.write(local_info), comm.read())
-        except Exception as e:
-            with suppress(Exception):
-                await comm.close()
-            raise CommClosedError(f"Comm {comm!r} closed.") from e
+        await comm.write(local_info)
+        handshake = await comm.read()
```
Comment (graingert) [inline distributed/comm/core.py:223]: why not use the gather trick here also?

```suggestion
        _, handshake = await asyncio.gather(comm.write(local_info), comm.read())
```

File: distributed/comm/core.py [hunk 2]
```diff
@@ -323,17 +310,7 @@ def time_left():
         **comm.handshake_info(),
         **(handshake_overrides or {}),
     }
-    try:
-        # This would be better, but connections leak if worker is closed quickly
-        # write, handshake = await asyncio.gather(comm.write(local_info), comm.read())
-        handshake = await wait_for(comm.read(), time_left())
-        await wait_for(comm.write(local_info), time_left())
```
[the remark this answers is not in the document]
Comment (graingert) [reply distributed/comm/core.py:330]: the tcp comm.write() doesn't actually wait for anything, it actually returns immediately:

the `async def write(` only has one await in it:

https://github.com/dask/distributed/blob/9bd90c041a4a10b649e7551ed6ceac89f9757507/distributed/comm/tcp.py#L271
which dispatches large objects to a thread pool (local_info is not large)

then it writes to the stream with:
https://github.com/dask/distributed/blob/9bd90c041a4a10b649e7551ed6ceac89f9757507/distributed/comm/tcp.py#L323
which does not wait for the buffer to drain (flow control)

[the remark this answers is not in the document]
Comment (graingert) [reply distributed/comm/core.py:330]: I think the connection leak issue was resolved here https://github.com/graingert/distributed/commit/b59a322756a125540c912522f6769eb3cb18b807

if `read()` or `write()` are cancelled the comm is aborted

**Review by graingert — APPROVED** [review]

File: distributed/tests/test_client.py [hunk 3]
```diff
@@ -5819,22 +5819,6 @@ async def test_client_timeout_2():
         assert stop - start < 1
 
 
-@gen_test()
-async def test_client_active_bad_port():
```
Comment (graingert) [inline distributed/tests/test_client.py:5823]: http servers wait for a `\r\n` before parsing the request line and so we never get a handshake back from those servers so this test fails

**Comment by fjetter** [comment]
✅  From my end. @graingert feel free to merge

File: distributed/worker.py [hunk 4]
```diff
@@ -1538,6 +1538,8 @@ async def close(  # type: ignore
         for pc in self.periodic_callbacks.values():
             pc.stop()
 
+        self.stop()
```
Comment (hendrikmakait) [inline distributed/worker.py:1541]: Moving this up here has broken P2P restarts by introducing a new race condition (or making it exponentially more likely to occur). Why do we need this change?

See https://github.com/dask/distributed/issues/8011#issuecomment-1656107174 for more on the P2P problem.

Comment (fjetter) [reply distributed/worker.py:1541]: I don't have any issues with reverting this change but I am a bit reluctant to revert the handshake timeout removal. I assume that this is required to keep many tests green.

**Pull request merged**